packages/tangle/tipmanager.go

```diff
@@ -235,20 +237,238 @@ func (t *TipManager) Tips(p payload.Payload, countParents int) (parents MessageI
+		for !t.tangle.Utils.AllTransactionsApprovedByMessages(transaction.ReferencedTransactionIDs(), parents...) || len(parents) == 0 {
 			if tries == 0 {
-				err = errors.Errorf("not able to make sure that all inputs are in the past cone of selected tips")
+				err = errors.Errorf("not able to make sure that all inputs are in the past cone of selected tips and parents have correct time-since-confirmation")
 				return nil, err
 			}
 			tries--
 
 			parents = t.selectTips(p, MaxParentsCount)
 		}
+	} else {
+		tries := 5
+		for len(parents) == 0 {
+			if tries == 0 {
+				err = errors.Errorf("not able to make sure that parents have correct time-since-confirmation")
+				return nil, err
+			}
+			tries--
+
+			parents = t.selectTips(p, MaxParentsCount)
+		}
+	}
+	return parents, nil
+}
+
+func (t *TipManager) isPastConeTimestampCorrect(messageID MessageID) (timestampValid bool) {
+	now := clock.SyncedTime()
+	markersVisited := 0
+	messagesVisited := 0
+	minSupportedTimestamp := now.Add(-5 * time.Minute)
+	timestampValid = true
+
+	if t.tangle.TimeManager.LastConfirmedMessage().MessageID == EmptyMessageID {
+		// if the genesis message is the last confirmed message, then there is no point in performing tangle walk
+		// return true so that the network can start issuing messages when the tangle starts
+		return
+	}
+
+	if t.tangle.TimeManager.LastConfirmedMessage().Time.Before(minSupportedTimestamp) {
+		return false
+	}
+
+	if t.tangle.ConfirmationOracle.IsMessageConfirmed(messageID) {
+		// selected message is confirmed, therefore it's correct
+		return
+	}
+
+	// selected message is not confirmed and older than TSC
+	t.tangle.Storage.Message(messageID).Consume(func(message *Message) {
+		timestampValid = message.IssuingTime().After(minSupportedTimestamp)
+	})
+	if !timestampValid {
+		// timestamp of the selected message is invalid
+		return
+	}
+
+	markerWalker := walker.New(false)
+	messageWalker := walker.New(false)
+
+	t.processMessage(messageID, messageWalker, markerWalker)
+
+	for markerWalker.HasNext() && timestampValid {
+		markersVisited++
+		timestampValid = t.checkMarker(markerWalker.Next().(markers.Marker), messageWalker, markerWalker, minSupportedTimestamp)
+	}
+
+	for messageWalker.HasNext() && timestampValid {
+		messagesVisited++
+		timestampValid = t.checkMessage(messageWalker.Next().(MessageID), messageWalker, minSupportedTimestamp)
 	}
+	return timestampValid
+}
 
+func (t *TipManager) processMessage(messageID MessageID, messageWalker, markerWalker *walker.Walker) {
+	t.tangle.Storage.MessageMetadata(messageID).Consume(func(messageMetadata *MessageMetadata) {
+		if messageMetadata.StructureDetails() == nil || messageMetadata.StructureDetails().PastMarkers.Size() == 0 {
+			// need to walk messages
+			messageWalker.Push(messageID)
+			return
+		}
+		messageMetadata.StructureDetails().PastMarkers.ForEach(func(sequenceID markers.SequenceID, index markers.Index) bool {
+			if sequenceID == 0 && index == 0 {
```

Comment:
When performing unit tests, it turns out that this sequence/index is returned sometimes when there are multiple messages approving the genesis. The first approver will get the `(0, 1)` marker, however the second, parallel approver will not get a new marker, so it's past marker is `(0, 0)` and in that case we need to walk its past cone. It's an unrealistic edge-case but it's covered, because otherwise there were some weird errors.